pkg/framework/pgdriver/driver.go

```diff
@@ -0,0 +1,103 @@
+package pgdriver
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/XSAM/otelsql"
+	"github.com/jackc/pgx/v5/pgxpool"
+	pgxstdlib "github.com/jackc/pgx/v5/stdlib"
+	"go.opentelemetry.io/otel/metric"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Option interface {
+	apply(*options)
+}
+
+type optionFunc func(c *options)
+
+func (fn optionFunc) apply(c *options) {
+	fn(c)
+}
+
+func WithTracerProvider(p trace.TracerProvider) Option {
+	return optionFunc(func(o *options) {
+		o.otelOptions = append(o.otelOptions, otelsql.WithTracerProvider(p))
+	})
+}
+
+func WithMeterProvider(p metric.MeterProvider) Option {
+	return optionFunc(func(o *options) {
+		o.otelOptions = append(o.otelOptions, otelsql.WithMeterProvider(p))
+	})
+}
+
+func WithContext(ctx context.Context) Option {
+	return optionFunc(func(o *options) {
+		o.ctx = ctx
+	})
+}
+
+type options struct {
+	connConfig  *pgxpool.Config
+	otelOptions []otelsql.Option
+	ctx         context.Context
+}
+
+type Driver struct {
+	pool *pgxpool.Pool
+	db   *sql.DB
+}
+
+func (d *Driver) DB() *sql.DB {
+	return d.db
+}
+
+func (d *Driver) Close() error {
+	d.pool.Close()
+
+	return nil
+}
+
+func NewPostgresDriver(url string, opts ...Option) (*Driver, error) {
+	config, err := pgxpool.ParseConfig(url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse postgres url: %w", err)
+	}
+
+	o := &options{
+		connConfig: config,
+		otelOptions: []otelsql.Option{
+			otelsql.WithAttributes(
+				semconv.DBSystemPostgreSQL,
+			),
+		},
+	}
+
+	for _, opt := range opts {
+		opt.apply(o)
+	}
+
+	if o.ctx == nil {
+		o.ctx = context.Background()
+	}
```

Comment:
I don't think anything is wrong with making this a mandatory argument.

Reply:
Yeah, I was a bit hesitant about having context.Context in constructorm but I agree that in this case it makes sense to enforce it.